File: src/stores/presets.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'
import type { Preset } from '../types'

export const usePresetsStore = defineStore('presets', () => {
  const presets = ref<Preset[]>([])

  async function fetchPresets() {
    try {
      presets.value = await invoke<Preset[]>('list_presets')
    } catch (e) {
      console.error('[presets] list_presets failed:', e)
      throw e
    }
    const hasDefault = presets.value.some(p => p.name === 'デフォルト')
    if (!hasDefault) {
      await createPreset({
        name: 'デフォルト',
        format: 'mp4',
        quality: 'best',
        output_dir: '~/Downloads/YTDown',
        embed_thumbnail: true,
        embed_metadata: true,
        write_subs: false,
        embed_subs: false,
        embed_chapters: true,
        sponsorblock: false,
      })
    }
  }

  async function createPreset(payload: Omit<Preset, 'id' | 'created_at'>): Promise<number> {
    try {
      const id = await invoke<number>('create_preset', {
        name: payload.name,
        format: payload.format,
        quality: payload.quality,
        outputDir: payload.output_dir,
        embedThumbnail: payload.embed_thumbnail,
        embedMetadata: payload.embed_metadata,
        writeSubs: payload.write_subs,
        embedSubs: payload.embed_subs,
        embedChapters: payload.embed_chapters,
        sponsorblock: payload.sponsorblock,
      })
      await fetchPresets()
      return id
    } catch (e) {
      console.error('[presets] create_preset failed:', e)
      throw e
    }
  }

  async function updatePreset(payload: Omit<Preset, 'created_at'>) {
    try {
      await invoke('update_preset', {
        id: payload.id,
        name: payload.name,
        format: payload.format,
        quality: payload.quality,
        outputDir: payload.output_dir,
        embedThumbnail: payload.embed_thumbnail,
        embedMetadata: payload.embed_metadata,
        writeSubs: payload.write_subs,
        embedSubs: payload.embed_subs,
        embedChapters: payload.embed_chapters,
        sponsorblock: payload.sponsorblock,
      })
      await fetchPresets()
    } catch (e) {
      console.error('[presets] update_preset failed:', e)
      throw e
    }
  }

  async function deletePreset(id: number) {
    try {
      await invoke('delete_preset', { id })
      await fetchPresets()
    } catch (e) {
      console.error('[presets] delete_preset failed:', e)
      throw e
    }
  }

  return { presets, fetchPresets, createPreset, updatePreset, deletePreset }
})

File: src-tauri/src/commands/presets.rs
use crate::db::{models::Preset, queries};
use crate::state::AppState;
use tauri::State;

#[tauri::command]
pub async fn list_presets(state: State<'_, AppState>) -> Result<Vec<Preset>, String> {
    let presets = {
        let db = state.db.lock().await;
        queries::list_presets(&db).map_err(|e| e.to_string())?
    };
    Ok(presets)
}

#[tauri::command]
pub async fn create_preset(
    name: String,
    format: String,
    quality: String,
    output_dir: String,
    embed_thumbnail: bool,
    embed_metadata: bool,
    write_subs: bool,
    embed_subs: bool,
    embed_chapters: bool,
    sponsorblock: bool,
    state: State<'_, AppState>,
) -> Result<i64, String> {
    let id = {
        let db = state.db.lock().await;
        queries::insert_preset(
            &db,
            &name,
            &format,
            &quality,
            &output_dir,
            embed_thumbnail,
            embed_metadata,
            write_subs,
            embed_subs,
            embed_chapters,
            sponsorblock,
        )
        .map_err(|e| e.to_string())?
    };
    Ok(id)
}

#[tauri::command]
pub async fn update_preset(
    id: i64,
    name: String,
    format: String,
    quality: String,
    output_dir: String,
    embed_thumbnail: bool,
    embed_metadata: bool,
    write_subs: bool,
    embed_subs: bool,
    embed_chapters: bool,
    sponsorblock: bool,
    state: State<'_, AppState>,
) -> Result<(), String> {
    {
        let db = state.db.lock().await;
        queries::update_preset(
            &db,
            id,
            &name,
            &format,
            &quality,
            &output_dir,
            embed_thumbnail,
            embed_metadata,
            write_subs,
            embed_subs,
            embed_chapters,
            sponsorblock,
        )
        .map_err(|e| e.to_string())?
    };
    Ok(())
}

#[tauri::command]
pub async fn delete_preset(id: i64, state: State<'_, AppState>) -> Result<(), String> {
    {
        let db = state.db.lock().await;
        queries::delete_preset(&db, id).map_err(|e| e.to_string())?
    };
    Ok(())
}
